File: strings/minimum_genetic_mutation.kt
fun main() {

    val start = "AACCGGTT"
    val end = "AAACGGTA"
    val bank = arrayOf<String>("AACCGGTA", "AACCGCTA", "AAACGGTA")

    println(minMutation(start, end, bank))
}

/*
    * Starting point is assumed to be valid, so it might not be included in the bank.
    * If multiple mutations are needed, all mutations during in the sequence must be valid.
    * You may assume start and end string is not the same.
*/

fun minMutation(start: String, end: String, bank: Array<String>): Int {
    if (end !in bank) {
        return -1
    }

    val allCombos = mutableMapOf<String, MutableList<String>>()

    // populate dictionary of wildcards
    for (word in bank) {
        for (i in start.indices) {
            val wordArray = word.toCharArray()
            wordArray[i] = '*'
            val newWord = String(wordArray)
            if (!allCombos.containsKey(newWord)) {
                allCombos[newWord] = mutableListOf<String>()
            }
            allCombos[newWord]!!.add(word)
        }
    }

    val queue = ArrayDeque<Pair<String, Int>>()
    queue.addFirst(Pair<String, Int>(start, 1))

    while (queue.isNotEmpty()) {
        val (currentWord, level) = queue.removeLast()
        for (i in start.indices) {
            val wordArray = currentWord.toCharArray()
            wordArray[i] = '*'
            val newWord = String(wordArray)
            if (!allCombos.containsKey(newWord)) continue
            val combos = allCombos[newWord]!!
            for (word in combos) {
                if (word == end) {
                    // we found it!
                    return level
                }
                queue.addFirst(Pair<String, Int>(word, level+1))
            }
            allCombos[newWord] = mutableListOf<String>()
        }
    }
    return -1
}

File: strings/minimum_genetic_mutation.py
#! /usr/bin/env python3

class Solution:
    def minMutation(self, start: str, end: str, bank: list[str]) -> int:
        if end not in bank:
            return -1

        combos = {}
        for word in bank:
            for i in range(len(word)):
                wild = word[:i] + "*" + word[i+1:]
                if wild not in combos:
                    combos[wild] = []
                combos[wild].append(word)

        queue = []
        queue.append((start, 1))
        while len(queue) > 0:
            word, count = queue.pop(0)
            for i in range(len(word)):
                wild = word[:i] + "*" + word[i+1:]
                if wild not in combos:
                    continue
                for w in combos[wild]:
                    if w == end:
                        return count
                    queue.append((w, count+1))
                del(combos[wild])

        return -1


if __name__ == "__main__":
    s = Solution()
    start = "AACCGGTT"
    end = "AAACGGTA"
    bank = ["AACCGGTA", "AACCGCTA", "AAACGGTA"]
    print(s.minMutation(start, end, bank))
